File: programmers/BridgeTruck.py
from collections import deque

def solution(bridge_len, weight, truck_wei):
    T = deque(truck_wei)
    B = deque([0] * bridge_len)
    time = 0
    now_wei = 0
    
    while T:
        time += 1
        now_wei -= B.popleft()
        if now_wei+T[0] <= weight:
            t = T.popleft()
            B.append(t)
            now_wei += t
        else:
            B.append(0)
    
    return time + bridge_len

''' 42583
# 다리를 지나는 트럭 - queue
    -> 시간의 흐름에 따라 반복문 통과 (트럭이 못올라가면 append(0))

    *반복문 안에서 sum() 쓰면 테스트케이스 하나 시간초과 -> now_wei 변수로 따로 관리
'''

File: programmers/BridgeTruck.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Deque<Integer> wait = new ArrayDeque<>();
        for (int t : truck_weights) wait.add(t);
        
        Deque<Integer> brid = new ArrayDeque<>();
        for (int i = 0; i < bridge_length; i++) brid.add(0);
        
        int time = 0;
        while (wait.size() > 0) {
            time++;
            brid.poll();
            
            int sum = brid.stream().mapToInt(Integer::intValue).sum();
            if (sum + wait.peek() <= weight) brid.add(wait.poll());
            else brid.add(0);
        }
        
        return time + bridge_length;
    }
}

/* 42583
# 다리를 지나는 트럭 - queue

 -> Queue<Integer> q = new ArrayDeque<>();로도 가능 : 기능보장(제한), 성능효율
    q.offer(e), poll, peek 사용
*/
